File: ios/backbone/LocalNotificationManager.m
//
//  LocalNotificationManager.m
//  Backbone
//
//  Created by Eko Mirhard on 10/2/16.
//  Copyright © 2016 Backbone Labs, Inc. All rights reserved.
//

#import "LocalNotificationManager.h"

#import "Constant.h"

@implementation LocalNotificationManager

/**
 Schedule local notifications based on the moduleName
 @param moduleName Name of the ActivityModule set as userInfo in the notification
 */
+ (BOOL)scheduleNotification:(NSString*)moduleName {
  if ([moduleName isEqualToString:@"posture"]) {
    UILocalNotification *localNotif = [[UILocalNotification alloc] init];
    if (localNotif) {
      localNotif.alertBody = NSLocalizedString(@"Your posture is not optimal!", nil);
      localNotif.soundName = UILocalNotificationDefaultSoundName;
      localNotif.userInfo = @{
                              @"module": moduleName
                              };
      
      [[UIApplication sharedApplication] scheduleLocalNotification:localNotif];
    }
    else {
      return NO;
    }
  }
  else if ([moduleName isEqualToString:@"step"]) {
    for (int i = 1; i <= NOTIFICATION_CYCLE / NOTIFICATION_PERIOD; i++) {
      UILocalNotification *newNotif = [[UILocalNotification alloc] init];
      
      if (newNotif) {
        newNotif.timeZone = [NSTimeZone localTimeZone];
        newNotif.fireDate = [NSDate dateWithTimeIntervalSinceNow:NOTIFICATION_PERIOD * 60.0 * i];
        newNotif.alertBody = NSLocalizedString(@"Go and take a walk!", nil);
        newNotif.soundName = UILocalNotificationDefaultSoundName;
        newNotif.repeatInterval = NSCalendarUnitHour;
        newNotif.userInfo = @{
                              @"module": moduleName
                              };
        
        [[UIApplication sharedApplication] scheduleLocalNotification:newNotif];
      }
      else {
        return NO;
      }
    }
  }
  
  return YES;
}

/**
 Check for any scheduled notification with the following moduleName set as userInfo
 @param moduleName Name of the ActivityModule set as userInfo in the notification
 */
+ (BOOL)hasScheduledNotification:(NSString *)moduleName {
  NSArray *eventArray = [[UIApplication sharedApplication] scheduledLocalNotifications];
  
  for (UILocalNotification *localNotif in eventArray) {
    NSDictionary *userInfo = localNotif.userInfo;
    
    if ([[userInfo objectForKey:@"module"] isEqualToString:moduleName]) {
      return YES;
    }
  }
  
  return NO;
}

/**
 Cancel all scheduled notification with the following moduleName set as userInfo
 @param moduleName Name of the ActivityModule set as userInfo in the notification
 */
+ (void)cancelScheduledNotification:(NSString *)moduleName {
  NSArray *eventArray = [[UIApplication sharedApplication] scheduledLocalNotifications];
  
  for (UILocalNotification *localNotif in eventArray) {
    NSDictionary *userInfo = localNotif.userInfo;
    
    if ([[userInfo objectForKey:@"module"] isEqualToString:moduleName]) {
      [[UIApplication sharedApplication] cancelLocalNotification:localNotif];
    }
  }
}

@end

File: android/app/src/main/java/co/backbonelabs/backbone/LocalNotificationManager.java
package co.backbonelabs.backbone;

import android.content.Context;
import android.os.Handler;
import android.util.Log;

import java.util.HashMap;

public class LocalNotificationManager {
    private static Context mContext;
    private static final String TAG = "LocalNotificationMgr";

    private static HashMap<String, Runnable> scheduledNotifications;

    private static final Integer STEP_NOTIFICATION_ID = 0x100;
    private static final Integer STEP_NOTIFICATION_PERIOD = 5; // in 'Minute'

    private static Handler notificationHandler;

    public LocalNotificationManager(Context context) {
        mContext = context;
        notificationHandler = new Handler();
        scheduledNotifications = new HashMap<String, Runnable>();
    }

    /**
     * Schedule local notifications based on the moduleName
     * @param moduleName Name of the ActivityModule set as userInfo in the notification
     */
    public static void scheduleNotification(String moduleName) {
        if (moduleName.equals("step")) {
            Runnable idleNotificationRunnable = new Runnable() {
                public void run() {
                    Log.d(TAG, "Idle TimeOut Notification");
                    NotificationService.sendNotification(STEP_NOTIFICATION_ID, "Go and take a walk!");
                    notificationHandler.postDelayed(this, STEP_NOTIFICATION_PERIOD * 1000 * 60);
                }
            };

            scheduledNotifications.put(moduleName, idleNotificationRunnable);
            notificationHandler.postDelayed(idleNotificationRunnable, STEP_NOTIFICATION_PERIOD * 1000 * 60);
        }
    }

    /**
     * Check for any scheduled notification with the following moduleName set as userInfo
     * @param moduleName Name of the ActivityModule set as userInfo in the notification
     * @return A boolean indicating whether any scheduled notification matching the criteria exists
     */
    public static boolean hasScheduledNotification(String moduleName) {
        return scheduledNotifications.containsKey(moduleName);
    }

    /**
     * Cancel all scheduled notification with the following moduleName set as userInfo
     * @param moduleName Name of the ActivityModule set as userInfo in the notification
     */
    public static void cancelScheduledNotification(String moduleName) {
        if (hasScheduledNotification(moduleName)) {
            notificationHandler.removeCallbacks(scheduledNotifications.get(moduleName));
            scheduledNotifications.remove(moduleName);
        }
    }
}
